osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs

```diff
@@ -268,6 +296,44 @@ private double computeFlashlightRating(double flashlightDifficultyValue, Mod[] m
             return flashlightRating * Math.Sqrt(ratingMultiplier);
         }
 
+        private static double calculateAimVisibilityFactor(double approachRate, double relevantMechanicalDifficultyRating)
+        {
+            const double ar_factor_end_point = 11.5;
+
+            double mechanicalDifficultyFactor = DifficultyCalculationUtils.ReverseLerp(relevantMechanicalDifficultyRating, 5, 10);
+            double arFactorStartingPoint = double.Lerp(9, 10.33, mechanicalDifficultyFactor);
+
+            return DifficultyCalculationUtils.ReverseLerp(approachRate, ar_factor_end_point, arFactorStartingPoint);
+        }
+
+        private static double calculateSpeedVisibilityFactor(double approachRate, double relevantMechanicalDifficultyRating)
+        {
+            const double ar_factor_end_point = 11.5;
+
+            double mechanicalDifficultyFactor = DifficultyCalculationUtils.ReverseLerp(relevantMechanicalDifficultyRating, 5, 10);
+            double arFactorStartingPoint = double.Lerp(10, 10.33, mechanicalDifficultyFactor);
+
+            return DifficultyCalculationUtils.ReverseLerp(approachRate, ar_factor_end_point, arFactorStartingPoint);
+        }
+
+        private static double calculateRelevantMechanicalDifficulty(double aimDifficultyValue, double speedDifficultyValue)
+        {
+            double aimValue = OsuStrainSkill.DifficultyToPerformance(Math.Sqrt(aimDifficultyValue) * difficulty_multiplier);
```

Comment:
All skills use `Math.Sqrt(difficultyValue) * difficulty_multiplier` right? We should move that out somewhere (which would also make it more obvious why we even do that)

Reply:
I would've wanted it to be moved in .DifficultyValue() itself but it will interrupt stuff like DifficultStrainCount calculation

Reply:
I think it should be a separate virtual method in `Skill`s

Reply:
I added `calculateDifficultyRating` in the diff calc itself, we could probably move it to `OsuStrainSkill` or something as a method but we already have `DifficultyToPerformance` and vice versa methods there and it felt like it would've got difficult to track... Maybe we could add it to `OsuStrainUtils`? I dunno

Reply:
Let's keep it as per tsunyoku's commit for now and refactor it in a separate PR